build: simplify by flattening the dependency graph

theuni opened the pull request:
These changes speed up my build (default config/options/targets) by roughly 10%. I suspect the difference may be more significant in other build configs.

Before:
> $ time cmake --build build -j24
> real	3m26.932s

After:
> $ time cmake --build build -j24
> real	3m7.556s

Generally they allow for jobservers (either `make -jX` or `ninja`) to be better utilized. This can be verified using `top` while building and looking at the number of compiles running at any given time before/after these changes. Before, it's easy to observe periods of stalling when only one or two compiles are happening. After these changes, the compiler process count should mostly match the number of jobs given (`-jX`) until it falls off at the end.

---

The first commit sets [DEPENDS_EXPLICIT_ONLY](https://cmake.org/cmake/help/latest/command/add_custom_command.html#command:add_custom_command) for commands which generate our test header files. Without this option, `test_bitcoin`'s generated headers won't be built until all of its other dependencies have been built. This introduces a significant stall in the build, though currently only Ninja benefits from this being set, and only CMake >= 3.27 understands it.

Example from a generated `build.ninja`:

Before:

> \# Custom command for src/test/data/base58_encode_decode.json.h
> 
> build src/test/data/base58_encode_decode.json.h | ${cmake_ninja_workdir}src/test/data/base58_encode_decode.json.h: CUSTOM_COMMAND /home/cory/dev/bitcoin/src/test/data/base58_encode_decode.json /home/cory/dev/bitcoin/cmake/script/GenerateHeaderFromJson.cmake || libcrc32c.a libcrc32c_sse42.a libleveldb.a libminisketch.a minisketch_clmul src/bitcoin_clientversion src/crypto/libbitcoin_crypto.a src/crypto/libbitcoin_crypto_avx2.a src/crypto/libbitcoin_crypto_sse41.a src/crypto/libbitcoin_crypto_x86_shani.a src/generate_build_info src/libbitcoin_cli.a src/libbitcoin_common.a src/libbitcoin_consensus.a src/libbitcoin_node.a src/secp256k1/src/libsecp256k1.a src/secp256k1/src/secp256k1_precomputed src/test/util/libtest_util.a src/univalue/libunivalue.a src/util/libbitcoin_util.a src/wallet/libbitcoin_wallet.a src/zmq/libbitcoin_zmq.a

After:

> \# Custom command for src/test/data/base58_encode_decode.json.h
> 
> build src/test/data/base58_encode_decode.json.h | ${cmake_ninja_workdir}src/test/data/base58_encode_decode.json.h: CUSTOM_COMMAND /home/cory/dev/bitcoin/src/test/data/base58_encode_decode.json /home/cory/dev/bitcoin/cmake/script/GenerateHeaderFromJson.cmake

---

The second commit is more significant. It sets [CMAKE_OPTIMIZE_DEPENDENCIES](https://cmake.org/cmake/help/latest/prop_tgt/OPTIMIZE_DEPENDENCIES.html) globally, which allows the objects of static libs to be built in parallel when one lib depends on the other. This can be set as a per-lib property, ~but I don't see any need for that as we don't currently have any edge-cases where this wouldn't be ok. If those should arise, we could always disable on a per-lib basis~.

Edit: turns out this triggers an [upstream bug](https://gitlab.kitware.com/cmake/cmake/-/issues/24058), which I guess can be considered an edge-case until fixed in CMake. I've added 2 per-lib opt-outs as a result.

Example:

Before:

> \# Link the static library src/libbitcoin_cli.a
> 
> build src/libbitcoin_cli.a: CXX_STATIC_LIBRARY_LINKER__bitcoin_cli_RelWithDebInfo src/CMakeFiles/bitcoin_cli.dir/compat/stdin.cpp.o src/CMakeFiles/bitcoin_cli.dir/rpc/client.cpp.o || src/univalue/libunivalue.a

After:

> \# Link the static library src/libbitcoin_cli.a
> 
> build src/libbitcoin_cli.a: CXX_STATIC_LIBRARY_LINKER__bitcoin_cli_RelWithDebInfo src/CMakeFiles/bitcoin_cli.dir/compat/stdin.cpp.o src/CMakeFiles/bitcoin_cli.dir/rpc/client.cpp.o
>

theuni committed:
build: avoid unnecessary dependencies on generated headers

This prevents the generation of these headers from also depending on the
dependencies of the libs/binaries which consume them.

Specifically, this prevents generated test headers (such as
test/data/base58_encode_decode.json.h) from depending on the
dependencies of test_bitcoin (libcrc32c.a libcrc32c_sse42.a libleveldb.a)

Note that this is currently only relevant for Ninja.

For more detail, see:
https://cmake.org/cmake/help/latest/command/add_custom_command.html

theuni committed:
build: simplify dependency graph

Allow the objects of static libs to be built in parallel rather than serially
based on their dependency ordering.

For more detail, see:
https://cmake.org/cmake/help/latest/prop_tgt/OPTIMIZE_DEPENDENCIES.html

theuni commented:
Looks like this is hitting a CMake bug: https://gitlab.kitware.com/cmake/cmake/-/issues/24058

Converting to draft while I investigate.

willcl-ark commented:
Concept ACK.

I noticed a bit of a "gap" in our compilation on master the other day."

![screenshot_20240916-221513](https://github.com/user-attachments/assets/8b16738b-858e-495b-bae1-33883efa1d02)

This PR appears to fix it quite nicely:

![image](https://github.com/user-attachments/assets/681e4105-a72a-4168-9af0-6097d753a17c)

I didn't see quite the same speedup as you though, but will investigate and do some more thorough testing/review tomorrow.

These tests were also using a default build, run using ninja with `-j16`.

hebasto reviewed (commented):
Concept ACK.

a39a915c226775396f6505efe77a4aba4d0ae3ad
Did you consider setting `CMAKE_ADD_CUSTOM_COMMAND_DEPENDS_EXPLICIT_ONLY` globally?

aaf30ef89465c18228b4e46fd1378b9087274cbe
> turns out this triggers an [upstream bug](https://gitlab.kitware.com/cmake/cmake/-/issues/24058), which I guess can be considered an edge-case until fixed in CMake. I've added 2 per-lib opt-outs as a result.

This bug may also affect the `crc32c` library after [switching](https://github.com/bitcoin/bitcoin/pull/30905) to the upstream build system.

hebasto commented on cmake/module/GenerateHeaders.cmake:
```diff
@@ -2,12 +2,19 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or https://opensource.org/license/mit/.
 
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.27)
+  set(DEPENDS_EXPLICIT_OPT DEPENDS_EXPLICIT_ONLY)
+else()
+  set(DEPENDS_EXPLICIT_OPT)
+endif()
+
```
A note for reviewers. The Guix builds are using CMake 3.24.2 (the `cmake-minimal` package).

hebasto commented:
> > ... though currently only Ninja benefits from this being set, and only CMake >= 3.27 understands it.
> 
> I wonder if one CI task should be using Ninja (and cmake >= 3.27), if it isn't too hard to implement. Otherwise this config will remain untested and errors may sneak in to the master branch, only being detected after merge.

Please consider https://github.com/bitcoin/bitcoin/pull/30915.

theuni commented:
> Concept ACK.
> 
> [a39a915](https://github.com/bitcoin/bitcoin/commit/a39a915c226775396f6505efe77a4aba4d0ae3ad) Did you consider setting `CMAKE_ADD_CUSTOM_COMMAND_DEPENDS_EXPLICIT_ONLY` globally?
> 
Yes, but I didn't feel as comfortable setting this one globally due to our (for ex) phony deploy targets. I figured we may make some changes in the future that could introduce hard-to-track-down bugs. It's clear to see that the header generation is self-contained, and afaik that's the only place that would matter for build performance.

willcl-ark commented:
ACK 42edb2f4a5900e0b2451a7446d58962c0bf0095d

I agree with @l0rinc that this seems conceptually nicer, and also see a much smaller time diff since 30888:

```log
# master @ 33adc7521cc8bb24b941d959022b084002ba7c60
________________________________________________________
Executed in  181.29 secs    fish           external
   usr time   38.26 mins  214.00 micros   38.26 mins
   sys time    1.49 mins   55.00 micros    1.49 mins


# this PR rebased on master @ 33adc7521cc8bb24b941d959022b084002ba7c60
________________________________________________________
Executed in  179.09 secs    fish           external
   usr time   39.35 mins  202.00 micros   39.35 mins
   sys time    1.47 mins   56.00 micros    1.47 mins
```

I also still see a "nicer" build process structure after this PR (using ninja tracing):

![image](https://github.com/user-attachments/assets/f1184fa6-e976-44dc-82e9-08f46a1fc88e)

hebasto reviewed (commented):
Tested and benchmarked on Ubuntu 23.10 with CMake 3.27.4 using the following command:
```
$ rm -rf build && cmake -B build -G Ninja --preset dev-mode -DWITH_MULTIPROCESS=OFF -DWITH_CCACHE=OFF && time cmake --build build -j 16
```

- The master branch @ 6fc4692797121b54de0c54e5b09ee47f322c038a:
```
real	6m44.237s
```

- 330d16e1aa8fe8d7a5bb755189d0d2991fef8a43:
```
real	6m45.264s
```

- 42edb2f4a5900e0b2451a7446d58962c0bf0095d
```
real	6m45.859s
```

I'm OK with the flattened dependency graph, but the expected build time improvement is questionable. At least, the "speed up" phrase should be removed from the PR description and commit messages.

edilmedeiros commented:
> I noticed a bit of a "gap" in our compilation on master the other day."

I have been noticing this "gap" all the time with cmake. I'll check with this PR.

theuni commented:
Updated to address hebasto's comment, and removed comments about it being a speedup. Should be good to go now.

This speeds things up significantly for me in some cases where I'm only making specific targets. I think it should be a general speedup in other cases too, but sure, it's more about the dependency reduction.

l0rinc commented on CMakeLists.txt:
```diff
@@ -71,6 +71,13 @@ set(CMAKE_CXX_EXTENSIONS OFF)
+# Without this, if libfoo.a depends on libbar.a, libfoo's objects can't begin
+# to be compiled until libbar.a has been created.
+if (NOT DEFINED CMAKE_OPTIMIZE_DEPENDENCIES)
+  set(CMAKE_OPTIMIZE_DEPENDENCIES TRUE)
```
I'm not sure I understand when we're using ON/OFF and when TRUE/FALSE, but I understood both are accepted, so it's just a nit

purpleKarrot replied:
`1`, `ON`, `YES`, `TRUE`, `Y`, or any non-zero number (including floating point numbers) are treated as true. See: [basic-expressions](https://cmake.org/cmake/help/latest/command/if.html#basic-expressions). Sticking to `ON`/`OFF` consistently avoids this confusion. But yeah, it is a nit.

l0rinc reviewed (approved):
utACK 12fa9511b5fba18e83f88b7b831906595bcf2116

Compared to previous version, `CMAKE_OPTIMIZE_DEPENDENCIES` was moved after [project](https://github.com/bitcoin/bitcoin/blob/master/CMakeLists.txt#L43-L48) and only assigned if user didn't provide any value.

<details>
<summary>Details</summary>

```patch
git diff HEAD a6efc4ca368661fb82136ea4d68f3043d058dc48 
diff --git a/CMakeLists.txt b/CMakeLists.txt
index edfb63926a..86d5ed6448 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,6 +40,11 @@ set(CMAKE_TRY_COMPILE_PLATFORM_VARIABLES
   CMAKE_CXX_LINK_EXECUTABLE
 )
 
+# Flatten static lib dependencies.
+# Without this, if libfoo.a depends on libbar.a, libfoo's objects can't begin
+# to be compiled until libbar.a has been created.
+set(CMAKE_OPTIMIZE_DEPENDENCIES TRUE)
+
 project(BitcoinCore
   VERSION ${CLIENT_VERSION_MAJOR}.${CLIENT_VERSION_MINOR}.${CLIENT_VERSION_BUILD}
   DESCRIPTION "Bitcoin client software"
@@ -71,13 +76,6 @@ set(CMAKE_CXX_EXTENSIONS OFF)
 
 list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/module)
 
-# Flatten static lib dependencies.
-# Without this, if libfoo.a depends on libbar.a, libfoo's objects can't begin
-# to be compiled until libbar.a has been created.
-if (NOT DEFINED CMAKE_OPTIMIZE_DEPENDENCIES)
-  set(CMAKE_OPTIMIZE_DEPENDENCIES TRUE)
-endif()
-
 #=============================
 # Configurable options
 #=============================
 ```

</details>

hebasto commented:
It seems I lack a deep understanding of CMake's underlying behaviour.

The comment in c4e498300c7e6b23dc464eca75a2bc9f86270dab states:
> Specifically, this prevents generated test headers (such as
test/data/base58_encode_decode.json.h) from depending on the
dependencies of test_bitcoin (libcrc32c.a libcrc32c_sse42.a libleveldb.a)
>
> Note that this is currently only relevant for Ninja.

And this aligns perfectly with the CMake [docs](https://cmake.org/cmake/help/latest/command/add_custom_command.html):
> Indicates that the command's `DEPENDS` argument represents all files required by the command and implicit dependencies are not required.
>
> Without this option, if any target uses the output of the custom command, CMake will consider that target's dependencies as implicit dependencies for the custom command in case this custom command requires files implicitly created by those targets.

However, I fail to trigger header regeneration in the following scenario on the master branch:
```
$ cmake --version | head -1
cmake version 3.30.3
$ cmake -G Ninja -B build
$ cmake --build build -t test_bitcoin
<apply a patch to modify some of test_bitcoin's dependencies>
$ cmake --build build -t test_bitcoin  # headers are expected to be regenerated, but they are not
```

What am I doing wrong?

theuni commented:
$ cmake -G Ninja -B build
$ cmake --build build -t test_bitcoin
$ touch src/test/data/bip341_wallet_vectors.json
$ cmake --build build -t test_bitcoin

`[2/5] cd /home/cory/dev/bitcoin/build/src/test && /usr/bin/cmake -DJSON_SOURCE_PATH=/home/cory/dev/bitcoin/src/test/data/bip341_wallet_vectors.json -DHEADER_PATH=/home/cory/dev/bitcoin/build/src/test/data/bip341_wallet_vectors.json.h -P /home/cory/dev/bitcoin/cmake/script/GenerateHeaderFromJson.cmake`

Why would the headers need to be regenerated unless the .json/.raw files themselves were touched?

hebasto commented:
> Why would the headers need to be regenerated unless the .json/.raw files themselves were touched?

The `.json`/`.raw` files are _explicit_ dependencies specified via the `DEPENDS` option. The `DEPENDS_EXPLICIT_ONLY` option changes the treatment of `_implicit_ dependencies.

theuni commented:
Without that commit, the headers must be generated late in the build, after a bunch of other libs have been built:

`base58_encode_decode.json /home/cory/dev/bitcoin/cmake/script/GenerateHeaderFromJson.cmake || libcrc32c.a libcrc32c_sse42.a libleveldb.a libminisketch.a minisketch_clmul src/bitcoin_clientversion src/crypto/libbitcoin_crypto.a src/crypto/libbitcoin_crypto_avx2.a src/crypto/libbitcoin_crypto_sse41.a src/crypto/libbitcoin_crypto_x86_shani.a src/generate_build_info src/libbitcoin_cli.a src/libbitcoin_common.a src/libbitcoin_consensus.a src/libbitcoin_node.a src/secp256k1/src/libsecp256k1.a src/secp256k1/src/secp256k1_precomputed src/test/util/libtest_util.a src/univalue/libunivalue.a src/util/libbitcoin_util.a src/wallet/libbitcoin_wallet.a src/zmq/libbitcoin_zmq.a`

See? The header now depends on all of those libs. Meaning that test_bitcoin can't start until (for ex) zmq has been built. That introduces a stall in the build eventually. Imagine if you're doing a -j50 build and all that remains to be built is test_bitcoin and src/zmq/libbitcoin_zmq.a. Without this change, they can't be built in parallel, so the tests wait on everything else.

theuni commented:
Actually, @hebasto , what guarantees that the headers are built before the `test_bitcoin` objects? It seems to work ok but I'm not sure why it doesn't race?

Seems the headers should be an interface library that test_bitcoin depends on?

purpleKarrot commented:
I have to dig deeper what "header generation" actually is in this context here. But generally, `add_custom_target` and `add_dependencies` are not useful for code generation. You just need a `add_custom_command` that generates the files and you use those files in an `add_library`/`add_executable`/`target_sources`. As long as the custom command and the library/executable are defined in the same directory, dependencies will be resolved.

Custom commands are only useful for things that are built explicitly, like `make documentation`.

theuni commented:
Thanks, yeah, ignore that question. You got it exactly right. We're covered by the generated headers being an output of an add_custom_command and being added to the target.

hebasto commented:
> Without that commit, the headers must be generated late in the build, after a bunch of other libs have been built:
> 
> `base58_encode_decode.json /home/cory/dev/bitcoin/cmake/script/GenerateHeaderFromJson.cmake || libcrc32c.a libcrc32c_sse42.a libleveldb.a libminisketch.a minisketch_clmul src/bitcoin_clientversion src/crypto/libbitcoin_crypto.a src/crypto/libbitcoin_crypto_avx2.a src/crypto/libbitcoin_crypto_sse41.a src/crypto/libbitcoin_crypto_x86_shani.a src/generate_build_info src/libbitcoin_cli.a src/libbitcoin_common.a src/libbitcoin_consensus.a src/libbitcoin_node.a src/secp256k1/src/libsecp256k1.a src/secp256k1/src/secp256k1_precomputed src/test/util/libtest_util.a src/univalue/libunivalue.a src/util/libbitcoin_util.a src/wallet/libbitcoin_wallet.a src/zmq/libbitcoin_zmq.a`
> 
> See? The header now depends on all of those libs. Meaning that test_bitcoin can't start until (for ex) zmq has been built. That introduces a stall in the build eventually. Imagine if you're doing a -j50 build and all that remains to be built is test_bitcoin and src/zmq/libbitcoin_zmq.a. Without this change, they can't be built in parallel, so the tests wait on everything else.

Thanks! I get it know. I've compared the resulting `ninja.build` files, and it's clear that the first commit effectively removes _order-only dependencies_ in header build edges.

hebasto reviewed (approved):
ACK 12fa9511b5fba18e83f88b7b831906595bcf2116.

fanquake commented on cmake/minisketch.cmake:
```diff
@@ -74,6 +74,9 @@ add_library(minisketch STATIC EXCLUDE_FROM_ALL
   ${PROJECT_SOURCE_DIR}/src/minisketch/src/fields/generic_8bytes.cpp
 )
 
+# Workaround for https://gitlab.kitware.com/cmake/cmake/-/issues/24058
```
No new movement here, but someone has offered another work around: https://gitlab.kitware.com/cmake/cmake/-/issues/24058#note_1624877

theuni replied:
@hebasto Looking at this again, is there any reason [this needs to be an object library](https://github.com/bitcoin/bitcoin/blob/master/cmake/minisketch.cmake#L43)? It triggers the bug that necessitates the workaround here. I'm wondering if rather than using our hack or the (better) workaround suggested upstream, if we should just make it an archive library and avoid the issue altogether?

hebasto replied:
> @hebasto Looking at this again, is there any reason [this needs to be an object library](https://github.com/bitcoin/bitcoin/blob/master/cmake/minisketch.cmake#L43)?

I don't thin so.

> It triggers the bug that necessitates the workaround here. I'm wondering if rather than using our hack or the (better) workaround suggested upstream, if we should just make it an archive library and avoid the issue altogether?

From [Professional CMake: A Practical Guide 20th Edition](https://discourse.cmake.org/t/professional-cmake-a-practical-guide-20th-edition/13394):
> Some developers may find object libraries more natural if coming from a background where non-CMake projects defined their targets based on sources or object files rather than a related set of static libraries. In general, however, where there is a choice, static libraries will typically be the more convenient choice in CMake projects.

hebasto replied:
@theuni 

> Looking at this again, is there any reason [this needs to be an object library](https://github.com/bitcoin/bitcoin/blob/master/cmake/minisketch.cmake#L43)? It triggers the bug that necessitates the workaround here. I'm wondering if rather than using our hack or the (better) workaround suggested upstream, if we should just make it an archive library and avoid the issue altogether?

Please see https://github.com/bitcoin/bitcoin/pull/31880.
